### ISSUE TITLE
GENERAL_NAME_print mishandled embedded nuls in IA5 string values

### ISSUE THREAD

**Issue by t-j-h** [openssl/openssl#8997]
See [https://github.com/openssl/openssl/blob/master/crypto/x509v3/v3_alt.c](url)

```
    case GEN_EMAIL:
        BIO_printf(out, "email:%s", gen->d.ia5->data);
        break;

    case GEN_DNS:
        BIO_printf(out, "DNS:%s", gen->d.ia5->data);
        break;

    case GEN_URI:
        BIO_printf(out, "URI:%s", gen->d.ia5->data);
        break;
```

Each of these instances should be using ASN1_STRING_print to print the value rather than printing the raw ia5 string data. If there is an embedded end-of-string (\x00) character the user will be unaware.

For code using GENERAL_NAME_print to handle the various formats and then post-processing for checking logic, the wrong handling will occur.

For users working with the command line applications, incorrect output will occur.

If we have embedded end-of-string values in the subject or issuer name, those are displayed correctly. For general ASN1_STRING handling we fold special characters to a period (.) which is another way of indicating there is more data. 

Reproduce the issue with a certificate created with a SAN with DNS and an embedded \x00 within the DNS name (yes this is an invalid usage - however we shouldn't silently truncate the output which may lead to unexpected handling).

**Comment by acheevbhagat** [comment]
Hello I'm new to the project and would like to contribute. Just to verify, the only change that needs to be made is to have each case use ASN1_STRING_print instead of BIO_printf?

**Comment by acheevbhagat** [comment]
I was looking through the documentation and it looks as if though ASN1_STRING_print is deprecated. Do you know which flag for ASN1_STRING_print_ex would make it behave like ASN1_STRING_print? I can't tell which one would be the correct one.

**Comment by mattcaswell** [comment]
Probably just ASN1_STRFLGS_ESC_CTRL?

**Comment by levitte** [comment]
Do use `ASN1_STRING_print`, for this reason:

```
: ; grep ASN1_STRING_print crypto/x509*/*
crypto/x509/t_req.c:                        || ASN1_STRING_print(bp,
crypto/x509v3/v3_admis.c:            || ASN1_STRING_print(bp, namingAuthority->namingAuthorityText) <= 0
crypto/x509v3/v3_admis.c:            || ASN1_STRING_print(bp, namingAuthority->namingAuthorityUrl) <= 0
crypto/x509v3/v3_admis.c:                    || ASN1_STRING_print(bp, pinfo->registrationNumber) <= 0
crypto/x509v3/v3_admis.c:                        || ASN1_STRING_print(bp, val) <= 0
crypto/x509v3/v3_prn.c:            ASN1_STRING_print(bp, X509_EXTENSION_get_data(ex));
crypto/x509v3/v3_sxnet.c:        ASN1_STRING_print(out, id->user);
```

Quite simply put, we haven't converted to use `ASN1_STRING_print_ex` in our own code, and doing so would require that the flag get passed around, all the way from the calling application.  Since we obviously don't have that mechanism in place yet (that would be a subject for another PR), I don't see a reason for diving into that because if *this* issue.

(a separate issue should be raised regarding our own use of deprecated `ASN1_STRING_print`)

**Comment by mspncp** [comment]
Fixed by #9101.
